File: backend/app/shared/geo.py
"""Pure geometry helpers — shared kernel (no business meaning, no deps).

Lives here because more than one slice needs great-circle distance: attendance
(geofence flags) and jobs (route distance → fuel estimate). Per the dependency
rules, ``shared/`` may be imported by any slice and imports nothing internal.
"""

from __future__ import annotations

from math import asin, cos, radians, sin, sqrt

EARTH_RADIUS_M = 6_371_000.0


def haversine_m(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """Great-circle distance in metres between two lat/lng points."""
    p1, p2 = radians(lat1), radians(lat2)
    dphi = radians(lat2 - lat1)
    dlambda = radians(lng2 - lng1)
    a = sin(dphi / 2) ** 2 + cos(p1) * cos(p2) * sin(dlambda / 2) ** 2
    return 2 * EARTH_RADIUS_M * asin(sqrt(a))

File: technician-app/src/lib/geo.ts
/**
 * Pure geometry — the mobile mirror of `backend/app/shared/geo.py`.
 *
 * Lives in `lib/` (not the attendance feature) because more than one feature
 * needs great-circle distance: attendance confirms geofence crossings client-
 * side, and jobs will want route distance for its fuel estimate — the same
 * reason the backend keeps it in a shared kernel. Dependency-free and pure.
 */

export const EARTH_RADIUS_M = 6_371_000;

const toRad = (deg: number): number => (deg * Math.PI) / 180;

/** Great-circle distance in metres between two lat/lng points. */
export function haversineM(lat1: number, lng1: number, lat2: number, lng2: number): number {
  const p1 = toRad(lat1);
  const p2 = toRad(lat2);
  const dPhi = toRad(lat2 - lat1);
  const dLambda = toRad(lng2 - lng1);
  const a =
    Math.sin(dPhi / 2) ** 2 + Math.cos(p1) * Math.cos(p2) * Math.sin(dLambda / 2) ** 2;
  return 2 * EARTH_RADIUS_M * Math.asin(Math.sqrt(a));
}
